gtsfm/frontend/verifier/oanet.py

```diff
@@ -0,0 +1,188 @@
+            camera_intrinsics_i1: intrinsics for image #i1.
+            camera_intrinsics_i2: intrinsics for image #i2.
+
+        Returns:
+            Estimated rotation i2Ri1, or None if it cannot be estimated.
+            Estimated unit translation i2Ui1, or None if it cannot be estimated.
+            Indices of verified correspondences, of shape (N, 2) with N <= N3. These are subset of match_indices.
+            Inlier ratio w.r.t. the estimated model, i.e. #ransac inliers / # putative matches.
+        """
+        if match_indices.shape[0] < self._min_matches:
+            logger.info("[OANet] Not enough correspondences for verification.")
+            return self._failure_result
+
+        if len(keypoints_i1) < self._min_matches:
+            return self._failure_result
+
+        # only matched keypoints are fed into the model.
+        matched_keypoints_i1 = keypoints_i1.coordinates[match_indices[:, 0]]
+        matched_keypoints_i2 = keypoints_i2.coordinates[match_indices[:, 1]]
+
+        # using OANet's custom normalization utility, obtain two (N,2) arrays
+        normalized_keypoints = [
+            torch.from_numpy(self.__normalize_kpts(matched_keypoints_i1).astype(np.float32)).to(self.device),
+            torch.from_numpy(self.__normalize_kpts(matched_keypoints_i2).astype(np.float32)).to(self.device),
+        ]
+
+        # (N,4) -> (1,1,N,4)
+        corr = torch.cat(normalized_keypoints, dim=-1)
+        corr = corr.unsqueeze(0).unsqueeze(0)
+
+        data = {}
+        data["xs"] = corr
+        data["sides"] = []  # using no side-information.
+
+        with torch.no_grad():
+            # obtain length-2 list back `y_hat`, each a Tensor of shape (1,N)
+            y_hat, _ = self.model(data)
+
+        y = y_hat[-1][0, :].cpu().numpy()
+        inlier_mask = y > self.default_config_.inlier_threshold
+
+        v_corr_idxs = match_indices[inlier_mask]
+        inlier_ratio_est_model = np.mean(inlier_mask)
+
+        if len(v_corr_idxs) < 8:
+            return self._failure_result
+
+        # Call F or E matrix verifier on OANet output.
+        i2Ri1, i2Ui1, _, _ = self.ransac.verify(
+            keypoints_i1=keypoints_i1,
+            keypoints_i2=keypoints_i2,
+            match_indices=v_corr_idxs,
+            camera_intrinsics_i1=camera_intrinsics_i1,
+            camera_intrinsics_i2=camera_intrinsics_i2,
+        )
+        return i2Ri1, i2Ui1, v_corr_idxs, inlier_ratio_est_model
+
+    def __normalize_kpts(self, kpts: np.ndarray) -> np.ndarray:
+        """Normalize keypoint coordinates to have zero-mean and a scale of sqrt(2).
+
+        Args:
+            kpts: array of shape (N,2) representing **unnormalized** 2d keypoint coordinates.
+
+        Returns:
+            nkpts: array of shape (N,2) representing **normalized** 2d keypoint coordinates.
+        """
+        x_mean = np.mean(kpts, axis=0)
+        dist = kpts - x_mean
+        meandist = np.linalg.norm(dist, axis=1).mean()
```

Comment:
what is meandist?

Reply:
It's the average norm of every keypoint coordinate (x,y).